server/src/main/java/com/google/coffeehouse/common/Book.java

```diff
@@ -0,0 +1,159 @@
+  public static final String ISBN_FIELD_NAME = "isbn";
+  public static final String TITLE_FIELD_NAME = "title";
+
+  /** The message on the IllegalArgumentException when there is no "title" key */
+  public static final String NO_VALID_TITLE_FROM_MAP = 
+      "No valid \"" + TITLE_FIELD_NAME + "\" key defined.";
+  private String title;
+  private String author;
+  private String isbn;
+  private String bookId;
+
+  private Book(Builder builder) {
+    this.title = builder.title;
+    this.author = builder.author;
+    this.isbn = builder.isbn;
+    this.bookId = builder.idGenerator.generateId();
+  }
+
+  /** Overloaded static factory, calls other fromMap with null IdentifierGenerator. */
+  public static Book fromMap(Map bookInfo) {
+    return Book.fromMap(bookInfo, null);
+  }
+
+  /** 
+   * Creates a {@link Book} object from a Map with the relevant parameters as keys.
+   * @param bookInfo the Map that contains the information used to construct the Book. At a
+   *     minimum this includes a {@code "title"} key that is mapped to a String that describes
+   *     the title of the Book to be created. Optional keys include a {@code "author"} key 
+   *     that is mapped to a String describing the author of the book, as well as an 
+   *     {@code "isbn"} key that is mapped to a String that is the isbn number of the book
+   * @param idGen the {@link IdentifierGenerator} used when constructing the Book. If
+   *     null, the default generator will be used
+   * @return the created Book
+   * @throws IllegalArgumentException if no valid {@code "title"} key is defined
+   */
+  public static Book fromMap(Map bookInfo, IdentifierGenerator idGen) {
+    String title = (String) bookInfo.getOrDefault(TITLE_FIELD_NAME, null);
+    if (title == null) {
+      throw new IllegalArgumentException(NO_VALID_TITLE_FROM_MAP);
+    }
+
+    Book.Builder bookBuilder = Book.newBuilder(title);
+    if (bookInfo.containsKey(AUTHOR_FIELD_NAME)) {
+      bookBuilder.setAuthor((String) bookInfo.get(AUTHOR_FIELD_NAME));
+    }
+    if (bookInfo.containsKey(ISBN_FIELD_NAME)) {
+      bookBuilder.setIsbn((String) bookInfo.get(ISBN_FIELD_NAME));
+    }
+    if (idGen != null) {
+      bookBuilder.setIdGenerator(idGen);
+    }
+    return bookBuilder.build();
+  }
+
+  public String getTitle() {
+    return title;
+  }
+
+  /** Returns an optional because the author may, or may not, exist. */
+  public Optional<String> getAuthor() {
+    return author == null ? Optional.empty() : Optional.of(author);
+  }
+
+  /** Returns an optional because the ISBN may, or may not, exist. */
+  public Optional<String> getIsbn() {
+    return isbn == null ? Optional.empty() : Optional.of(isbn);
+  }
+
+  public String getBookId() {
+    return bookId;
+  }
+
+  public void setTitle(String title) {
+    this.title = title;
+  }
+
+  public void setAuthor(String author) {
+    this.author = author;
+  }
+
+  public void setIsbn(String isbn) {
```

Comment:
Can we confirm the type of the ISBN? I believe the database is treating it as a long, currently.

Reply:
It is currently a long.

Reply:
Just an update, talked with Lina and we'll have it be a string so that we can keep the '-' separations between parts of the ISBN.